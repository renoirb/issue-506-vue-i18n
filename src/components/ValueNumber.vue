<script>
export default {
  name: 'ValueNumber',
  props: {
    value: {
      type: Number,
      default: 0,
    },
    formattedAs: {
      type: String,
      default: 'locale',
    },
  },
  computed: {
    formatted() {
      const locale = this.$i18n.locale
      const value = this.value
      const formattedAs = String(this.formattedAs).toLowerCase()
      if (formattedAs === 'int') {
        // Maybe we want just an integer, even though we might have a float
        return Number(value).toFixed(0)
      } else if (formattedAs === 'string') {
        return String(value)
      } else if (formattedAs === 'locale') {
        /**
         * Format number according to ECMA and ISO conventions.
         * https://www.ecma-international.org/ecma-402/1.0/#sec-11.3.2
         * http://cldr.unicode.org/
         */
        let formatter
        if (this.normalizedNumberFormatOptions !== null) {
          formatter = new Intl.NumberFormat(
            locale,
            this.normalizedNumberFormatOptions
          )
        } else {
          formatter = new Intl.NumberFormat(locale)
        }
        return formatter.format(Number(value))
      }

      return value
    },
  },
  render(createElement) {
    const value = this.value
    const locale = this.$i18n.locale

    const formatted = this.formatted
    const formattedAs = String(this.formattedAs).toLowerCase()

    const attrs = {
      'data-value': String(value),
      'data-value-component': 'number',
      'data-formatted-as': formattedAs,
      'data-locale': locale,
    }
    if (this.normalizedNumberFormatOptions !== null) {
      attrs['data-number-format-options'] = JSON.stringify(
        this.normalizedNumberFormatOptions
      )
    }

    const slots = Object.keys(this.$slots).reduce(
      (arr, key) => arr.concat(this.$slots[key]),
      []
    )
    slots.push(formatted)
    return createElement(
      'span',
      {
        attrs: {
          ...this.$attrs,
          ...attrs,
        },
        props: this.$props,
        on: this.$listeners,
      },
      slots
    )
  },
}
</script>
