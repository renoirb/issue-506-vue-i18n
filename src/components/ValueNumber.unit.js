import ValueNumber from '@/components/ValueNumber'

const SampleComponent = {
  template: '<div><strong v-t="label" /><value-number :value="value" /></div>',
  components: { ValueNumber },
  props: {
    label: {
      type: String,
      default: 'Updates',
    },
    value: {
      type: Number,
      required: true,
    },
  },
}

describe('ValueNumber', () => {
  it('Exports a valid component', () => {
    expect(ValueNumber).toBeAComponent()
  })

  it('Accept locale as prop', async () => {
    let locale = 'de-DE'

    const { localVue, i18n } = createComponentMocks({
      i18n: {
        locale,
      },
    })

    const value = 10000

    const wrapper = mount(ValueNumber, {
      propsData: { value, locale },
      localVue,
      i18n,
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toBe('10.000') // !!!!!
    expect(wrapper.vm.value).toBe(value)

    locale = 'fr-CA'
    i18n.locale = locale
    await wrapper.vm.$nextTick()
    wrapper.setProps({ value, locale })
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Is possible to use in cunjunction with vue-i18n text translation', async () => {
    let locale = 'de-DE'

    const { localVue, i18n } = createComponentMocks({
      i18n: {
        locale,
      },
    })

    const value = 10000

    const wrapper = mount(SampleComponent, {
      propsData: { value, locale },
      localVue,
      i18n,
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.vm.value).toBe(value)
  })
})
