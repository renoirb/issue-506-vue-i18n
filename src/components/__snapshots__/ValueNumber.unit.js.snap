// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueNumber Accept locale as prop 1`] = `<span locale="de-DE" data-value="10000" data-value-component="number" data-formatted-as="locale" data-locale="de-DE">10,000</span>`;

exports[`ValueNumber Accept locale as prop 2`] = `<span locale="fr-CA" data-value="10000" data-value-component="number" data-formatted-as="locale" data-locale="fr-CA">10,000</span>`;

exports[`ValueNumber Is possible to use in cunjunction with vue-i18n text translation 1`] = `<div locale="de-DE"><strong>Updates</strong><span data-value="10000" data-value-component="number" data-formatted-as="locale" data-locale="de-DE">10,000</span></div>`;
